aesd-char-driver/main.c

```diff
@@ -0,0 +1,202 @@
+/**
+ * @file aesdchar.c
+ * @brief Functions and data related to the AESD char driver implementation
+ *
+ * Based on the implementation of the "scull" device driver, found in
+ * Linux Device Drivers example code.
+ *
+ * @author Dan Walkes
+ * @date 2019-10-22
+ * @copyright Copyright (c) 2019
+ *
+ */
+
+#include <linux/module.h>
+#include <linux/init.h>
+#include <linux/printk.h>
+#include <linux/types.h>
+#include <linux/cdev.h>
+#include <linux/slab.h> // kmalloc
+#include <linux/fs.h> // file_operations
+#include "aesdchar.h"
+#include "aesd-circular-buffer.h"
+int aesd_major =   0; // use dynamic major
+int aesd_minor =   0;
+
+MODULE_AUTHOR("Ashwin Ravindra"); /** TODO: fill in your name **/
+MODULE_LICENSE("Dual BSD/GPL");
+
+struct aesd_dev aesd_device;
+char* final_buffptr = NULL;
+
+int aesd_open(struct inode *inode, struct file *filp) {
+    PDEBUG("open");
+    struct aesd_dev *dev;
+    dev = container_of(inode->i_cdev, struct aesd_dev, cdev);
+    filp->private_data = dev;
+    return 0;
+}
+
+int aesd_release(struct inode *inode, struct file *filp) {
+    PDEBUG("release");
+    /*Nothing to handle here. aesd_cleanup_module is responsible for cleanup*/
+    return 0;
+}
+
+ssize_t aesd_read(struct file *filp, char __user *buf, size_t count, loff_t *f_pos) {
+    ssize_t retval = 0;
+    PDEBUG("read %zu bytes with offset %lld",count,*f_pos);
+    size_t* f_pos_offset = kmalloc(sizeof(size_t), GFP_KERNEL);
+    struct aesd_buffer_entry* read_entry;
+
+    /*Lock the buffer and find the entry to read*/
+    mutex_lock(&aesd_device.write_lock); 
+
+    /*If the entry is found, read the entry and copy to user*/
+    if((read_entry = aesd_circular_buffer_find_entry_offset_for_fpos(aesd_device.command_buffer, (size_t)*f_pos, f_pos_offset)) != NULL) {
+        ssize_t bytes_read = read_entry->size - *f_pos_offset;
+        if(bytes_read > count)
+            bytes_read = count;
+        copy_to_user(buf, read_entry->buffptr + *f_pos_offset, bytes_read);
+        *f_pos += bytes_read;
+        retval = bytes_read;
+    }
+
+    /*Unlock the buffer*/
+    mutex_unlock(&aesd_device.write_lock);
+    return retval;
+}
+
+ssize_t aesd_write(struct file *filp, const char __user *buf, size_t count, loff_t *f_pos) {
+    PDEBUG("write %zu bytes with offset %lld",count,*f_pos);
+    ssize_t retval = -ENOMEM;
+    static size_t final_count = 0;
+
+    /*Temporary buffer to store the string from user*/
+    char* temp_buffptr = kmalloc(count, GFP_KERNEL);
+    copy_from_user(temp_buffptr, buf, count);
+
+    /*If newline character not found, append to final_buffptr*/
+    if(strchr(temp_buffptr, '\n') == NULL) {
```

Comment:
@ashwinhr186 I don't think you are guaranteed null terminated strings, so this code may not always work as expected.